src/main/java/eu/dissco/core/digitalspecimenprocessor/service/ProcessingService.java

```diff
@@ -399,29 +423,62 @@ private void rollbackNewSpecimen(DigitalSpecimenRecord digitalSpecimenRecord,
     try {
       kafkaService.deadLetterEvent(
           new DigitalSpecimenEvent(enrichments, digitalSpecimenRecord.digitalSpecimen()));
     } catch (JsonProcessingException e) {
-      log.error("Fatal exception, unable to dead letter queue: " + digitalSpecimenRecord.id(), e);
+      log.error(DLQ_FAILED + digitalSpecimenRecord.id(), e);
     }
   }
 
-  private DigitalSpecimenRecord mapToDigitalSpecimenRecord(DigitalSpecimenEvent event) {
+  private void rollbackHandleCreation(List<DigitalSpecimenRecord> records){
+    var request = fdoRecordBuilder.buildRollbackCreationRequest(records);
     try {
-      return new DigitalSpecimenRecord(
-          fdoRecordBuilder.createNewHandle(event.digitalSpecimen()),
-          midsService.calculateMids(event.digitalSpecimen()),
-          1,
-          Instant.now(),
-          event.digitalSpecimen()
-      );
-    } catch (TransformerException | PidCreationException e) {
-      log.error("Failed to process record with id: {}",
-          event.digitalSpecimen().physicalSpecimenId(),
-          e);
+      handleComponent.rollbackHandleCreation(request);
+    } catch (PidCreationException | PidAuthenticationException e) {
+      var ids = records.stream().map(DigitalSpecimenRecord::id).toList();
+      log.error("Unable to rollback handles for new specimens. Bad handles: {}", ids);
+    }
+  }
+
+  private void filterUpdatesAndRollbackHandles(List<UpdatedDigitalSpecimenTuple> records){
+    var recordsToRollback = records.stream()
+        .filter(r -> fdoRecordBuilder.handleNeedsUpdate(r.currentSpecimen().digitalSpecimen(),
+            r.currentSpecimen().digitalSpecimen()))
+        .map(UpdatedDigitalSpecimenTuple::currentSpecimen)
+        .toList();
+
+    rollbackHandleUpdate(recordsToRollback);
+  }
+
+  private void rollbackHandleUpdate(List<DigitalSpecimenRecord> recordsToRollback){
+    try {
+      var request = fdoRecordBuilder.buildRollbackUpdateRequest(recordsToRollback);
+      handleComponent.rollbackHandleUpdate(request);
+    } catch (PidCreationException | PidAuthenticationException e) {
+      var ids = recordsToRollback.stream().map(DigitalSpecimenRecord::id).toList();
+      log.error("Unable to rollback handles for new specimens. Bad handles: {}", ids);
+    }
+  }
+
+  private DigitalSpecimenRecord mapToDigitalSpecimenRecord(DigitalSpecimenEvent event,
+      Map<String, String> pidMap) {
+    var handle = matchPidToDs(pidMap, event.digitalSpecimen().physicalSpecimenId());
+    if (handle==null){
       return null;
     }
+
+    return new DigitalSpecimenRecord(
+        matchPidToDs(pidMap, event.digitalSpecimen().physicalSpecimenId()),
+        midsService.calculateMids(event.digitalSpecimen()),
+        1,
+        Instant.now(),
+        event.digitalSpecimen()
+    );
   }
+
+  private String matchPidToDs(Map<String, String> pidMap, String physicalSpecimenId) {
+    return pidMap.get(physicalSpecimenId);
```

Comment:
I would put this one method call in the parent function

Reply:
Done